release-21.1: changefeedccl: implement pushback mechanism in changefeeds

miretskiy opened the pull request:
/cc @cockroachdb/release

Note: This is a partial backport of #66503.
The only code picked up from that PR is the addition of blocking buffer.
Instead of refactoring the code as done by the original PR, the blocking
buffer definition is added directly to the buffer.go -- as it existed
before refactoring.

The use of the pushback mechanism is gated by the
`changefeed.mem.pushback_enabled` setting (false by default).

The partial backport should be a no-op change (without the setting change).

Implement pushback in changefeeds.

These PRs were authored by @ajwerner .

Informs #66534

Release note (enterprise change): Improved the internal bufferring of
CHANGEFEEDs to be robust to bursts in traffic which exceed the
throughput
to the sink.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67205)
<!-- Reviewable:end -->

Commit:
release-21.1: changefeedccl: implement pushback mechanism in changefeeds

/cc @cockroachdb/release

Note: This is a partial backport of #66503.
The only code picked up from that PR is the addition of blocking buffer.
Instead of refactoring the code as done by the original PR, the blocking
buffer definition is added directly to the buffer.go -- as it existed
before refactoring.

The use of the pushback mechanism is gated by the
changefeed.mem.pushback_enabled setting (false by default).

The partial backport should be a no-op change (without the setting
change).

Implement pushback in changefeeds.

These PRs were authored by @ajwerner .

Informs #66534

Release note (enterprise change): Improved the internal bufferring of
CHANGEFEEDs to be robust to bursts in traffic which exceed the
throughput
to the sink.

HonoreDB reviewed (approved):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67205)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->
